File: app/src/main/java/com/saitejajanjirala/imdbclone/ui/search/SearchFragment.java
package com.saitejajanjirala.imdbclone.ui.search;

import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.navigation.NavDirections;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.saitejajanjirala.imdbclone.BaseApplication;
import com.saitejajanjirala.imdbclone.R;
import com.saitejajanjirala.imdbclone.adapters.SearchAdapter;
import com.saitejajanjirala.imdbclone.databinding.FragmentSearchBinding;
import com.saitejajanjirala.imdbclone.di.component.DaggerFragmentComponent;
import com.saitejajanjirala.imdbclone.di.module.FragmentModule;
import com.saitejajanjirala.imdbclone.models.Result;
import com.saitejajanjirala.imdbclone.models.SearchResult;
import com.saitejajanjirala.imdbclone.ui.home.HomeFragmentDirections;
import com.saitejajanjirala.imdbclone.utils.RxSearchObservable;

import java.util.ArrayList;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.ObservableSource;
import io.reactivex.rxjava3.functions.Consumer;
import io.reactivex.rxjava3.functions.Function;
import io.reactivex.rxjava3.functions.Predicate;
import io.reactivex.rxjava3.schedulers.Schedulers;


public class SearchFragment extends Fragment implements SearchAdapter.onSearchResultClickListener {

    FragmentSearchBinding binding;
    @Inject
    SearchViewModel searchViewModel;
    ArrayList<Result> results;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        DaggerFragmentComponent.builder()
                .applicationComponent(
                        ((BaseApplication) (requireContext().getApplicationContext())).applicationComponent)
                .fragmentModule(new FragmentModule(this))
                .build()
                .inject(this);
        results = new ArrayList<>();
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentSearchBinding.bind(inflater.inflate(R.layout.fragment_search, container, false));
        bindView();
        return binding.getRoot();
    }

    private void bindView() {
        LinearLayoutManager lm = new LinearLayoutManager(requireContext(), LinearLayoutManager.VERTICAL, false);
        binding.searchResults.setLayoutManager(lm);
        SearchAdapter adapter = new SearchAdapter(this, results);
        binding.searchResults.setAdapter(adapter);
        RecyclerView.ItemDecoration itemDecoration=new DividerItemDecoration(requireContext(), LinearLayout.VERTICAL);
        searchViewModel.searchResultMutableLiveData.observe(getViewLifecycleOwner(), new Observer<SearchResult>() {
            @Override
            public void onChanged(SearchResult searchResult) {
                if (searchResult != null) {
                    if(searchResult.results.isEmpty()){
                        binding.searchResults.setVisibility(View.GONE);
                        binding.noResultsLayout.setVisibility(View.VISIBLE);
                    }
                    else{
                        binding.searchResults.setVisibility(View.VISIBLE);
                        binding.noResultsLayout.setVisibility(View.GONE);
                    }
                    results.clear();
                    results.addAll(searchResult.results);
                    adapter.notifyDataSetChanged();
                }
            }
        });
        binding.backButton.setOnClickListener(v -> {
            closeFragment();
        });
        RxSearchObservable.fromView(binding.searchView)
                .debounce(300, TimeUnit.MILLISECONDS)
                .filter(new Predicate<String>() {
                    @Override
                    public boolean test(String s) throws Throwable {
                        return !TextUtils.isEmpty(s);
                    }
                })
                .distinctUntilChanged()
                .switchMap(new Function<String, ObservableSource<?>>() {
                    @Override
                    public ObservableSource<?> apply(String s) throws Throwable {
                        searchViewModel.getData(s);
                        return Observable.just("called");
                    }
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<Object>() {
                    @Override
                    public void accept(Object o) throws Throwable {

                    }
                });

    }

    @Override
    public void onResultClick(Result result) {
        NavDirections action=SearchFragmentDirections.actionSearchFragmentToDetailFragment(result);
        Navigation.findNavController(binding.getRoot()).navigate(action);
    }

    private void closeFragment() {
        Navigation.findNavController(binding.getRoot()).popBackStack();
    }

}

File: app/src/main/java/com/saitejajanjirala/imdbclonekotlin/ui/search/SearchFragment.kt
package com.saitejajanjirala.imdbclonekotlin.ui.search

import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.SearchView
import android.widget.Toast
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.NavDirections
import androidx.navigation.Navigation
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView.ItemDecoration
import com.saitejajanjirala.imdbclonekotlin.R
import com.saitejajanjirala.imdbclonekotlin.adapters.SearchAdapter
import com.saitejajanjirala.imdbclonekotlin.databinding.FragmentSearchBinding
import com.saitejajanjirala.imdbclonekotlin.models.Result
import dagger.hilt.android.AndroidEntryPoint
import java.util.*

@AndroidEntryPoint
class SearchFragment : Fragment(),
    SearchAdapter.onSearchResultClickListener {
    private lateinit var binding: FragmentSearchBinding
    private val searchViewModel: SearchViewModel by viewModels()
    private lateinit var results: ArrayList<Result>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        results = ArrayList()
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding =
            FragmentSearchBinding.bind(inflater.inflate(R.layout.fragment_search, container, false))
        bindView()
        return binding.root
    }

    private fun bindView() {
        val lm = LinearLayoutManager(requireContext(), LinearLayoutManager.VERTICAL, false)
        binding.searchResults.layoutManager = lm
        val adapter = SearchAdapter(this, results)
        binding.searchResults.adapter = adapter
        val itemDecoration: ItemDecoration =
            DividerItemDecoration(requireContext(), LinearLayout.VERTICAL)
        searchViewModel.searchResultMutableLiveData.observe(viewLifecycleOwner, { searchResult ->
            if (searchResult != null) {
                if (searchResult.results!=null && searchResult.results!!.isEmpty()) {
                    binding.searchResults.visibility = View.GONE
                    binding.noResultsLayout.visibility = View.VISIBLE
                } else {
                    binding.searchResults.visibility = View.VISIBLE
                    binding.noResultsLayout.visibility = View.GONE
                }
                results.clear()
                searchResult.results?.let { results.addAll(it) }
                adapter.notifyDataSetChanged()
            }

        })
        binding.backButton.setOnClickListener { v -> closeFragment() }
        binding.searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(p0: String?): Boolean {
                binding.searchView.clearFocus()
                if(p0!=null && !TextUtils.isEmpty(p0)) {
                    searchViewModel.getSearchResults(p0)
                }
                else{
                    Toast.makeText(requireContext(),"Query Cannot be empty",Toast.LENGTH_SHORT).show()
                }
                return true
            }

            override fun onQueryTextChange(p0: String?): Boolean {
                return false
            }
        })
        searchViewModel.msg.observe(viewLifecycleOwner,{
            if(it!=null && !TextUtils.isEmpty(it)){
                Toast.makeText(requireContext(),it,Toast.LENGTH_SHORT).show()
            }
        })
        searchViewModel.loading.observe(viewLifecycleOwner,{
            binding.progressbar.isVisible=it!=null && it
        })
    }

    override fun onResultClick(result: Result?) {
        if(result!=null) {
            val action: NavDirections =
                SearchFragmentDirections.actionSearchFragmentToDetailFragment(result)
            Navigation.findNavController(binding.root).navigate(action)
        }
    }

    private fun closeFragment() {
        Navigation.findNavController(binding.root).popBackStack()
    }
}